### PR TITLE
ZIO STM: add when & whenM (#2606)

### LINKED ISSUE #2606: STM: Add when & whenM
I think it would be useful for api parity with ZIO

### PR DESCRIPTION
**Pull request by KadekM**
Closes #2606

### CODE REVIEW

File: core/shared/src/main/scala/zio/stm/ZSTM.scala [hunk 1]
```diff
@@ -157,6 +157,16 @@ final class ZSTM[-R, +E, +A] private[stm] (
   def fallback[A1 >: A](a: => A1)(implicit ev: CanFail[E]): ZSTM[R, Nothing, A1] =
     fold(_ => a, identity)
 
+  /**
+   * The moral equivalent of `if (p) exp`
+   */
+  def when[R1 <: R, E1 >: E](b: Boolean): ZSTM[R1, E1, Unit] = ZSTM.when(b)(self)
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/stm/ZSTM.scala]: We don't need the additional type parameters here since they aren't used anywhere.

Comment (KadekM) [reply core/shared/src/main/scala/zio/stm/ZSTM.scala]: I wondered, but we also have them in https://github.com/zio/zio/blob/master/core/shared/src/main/scala/zio/ZIO.scala#L1559 and other places... should I also remove them there?

**Comment by KadekM** [comment]
@adamgfraser alphabetized and added tests

File: core/shared/src/main/scala/zio/stm/ZSTM.scala (hunk 1 above)
Comment (KadekM) [reply core/shared/src/main/scala/zio/stm/ZSTM.scala]: removed

File: core-tests/shared/src/test/scala/zio/stm/ZSTMSpec.scala [hunk 2]
```diff
@@ -414,6 +414,21 @@ object ZSTMSpec extends ZIOBaseSpec {
         } yield assert(result)(equalTo(2 -> 2))
       }
     ),
+    suite("when combinators")(
```
Comment (adamgfraser) [inline core-tests/shared/src/test/scala/zio/stm/ZSTMSpec.scala]: Can we also test that the transaction is not executed if the condition is false?

Comment (KadekM) [reply core-tests/shared/src/test/scala/zio/stm/ZSTMSpec.scala]: I've added them

**Comment by KadekM** [comment]
> Looks great. Will merge when it goes green. Also generally no need to force push. We squash commits when we merge and it generally makes it easier to review with the commits unless there is a specific reason you want to force it. Thanks for your contribution!

ah ok, will keep that in mind next time, generally I dont like to see 20 commits per 5lines of change :dancer: , but if you squash thats fine

**Pull request merged**